lib/semantics/check-do-concurrent.cc

```diff
@@ -160,25 +192,38 @@ class DoConcurrentEnforcement {
     }
     return false;
   }
+  bool addName(const std::optional<parser::Name> &nm) {
+    if (nm.has_value())
+      names_.insert(nm.value().ToString());
+    return true;
+  }
 
+  bool inDoConcurrent{false};
+  std::set<std::string> names_;
   std::set<parser::Label> labels_;
   parser::CharBlock currentStatementSourcePosition_;
   parser::Messages &messages_;
 };
 
 class DoConcurrentLabelEnforce {
 public:
-  DoConcurrentLabelEnforce(
-      parser::Messages &messages, std::set<parser::Label> &&labels)
-    : messages_{messages}, labels_{labels} {}
+  DoConcurrentLabelEnforce(parser::Messages &messages,
+                           std::set<parser::Label> &&labels,
+                           std::set<std::string> &&names)
+    : messages_{messages}, labels_{labels}, names_{names} {}
   template<typename T> bool Pre(const T &) { return true; }
   template<typename T> bool Pre(const parser::Statement<T> &statement) {
     currentStatementSourcePosition_ = statement.source;
     return true;
   }
+  bool Pre(const parser::DoConstruct &d) {
+    do_depth_++;
```

Comment:
It's a very minor point, but it's somewhat more idiomatic to use pre-increment (`++do_depth_`) in modern C++.  (And pre-decrement below.)

Reply:
OK. done.